h0rton/h0_inference/h0_posterior.py

```diff
@@ -0,0 +1,269 @@
+
+        """
+        # Seed for reproducibility
+        random.seed(seed)
+        np.random.seed(seed)
+        # Initialize output dict
+        samples_dict = dict(
+                            H0_samples=np.zeros(n_samples),
+                            H0_weights=np.zeros(n_samples),
+                            inferred_vd=np.zeros(n_samples),
+                            ll_vd=np.zeros(n_samples),
+                            inferred_td=np.zeros((n_samples, self.n_img)),
+                            ll_td=np.zeros((n_samples, self.n_img)),
+                            )
+
+        for n in tqdm(range(n_samples)):
+            H0_candidate = self.sample_H0()
+            k_ext = self.sample_kappa_ext()
+            aniso_param = self.sample_aniso_param()
+            # Define cosmology
+            cosmo = FlatLambdaCDM(H0=H0_candidate, Om0=self.Om0, Ob0=self.Ob0)
+            # Tool for getting time delays and velocity dispersions
+            lens_prop = LensProp(self.z_lens, self.z_src, self.kwargs_model, cosmo=cosmo)
+            # Velocity dispersion
+            if exclude_vel_disp:
+                ll_vd = 0.0
+            else:
+                inferred_vd = lens_prop.velocity_dispersion(self.kwargs_lens, r_eff=self.lens_light_R_sersic, R_slit=self.R_slit, dR_slit=self.dR_slit, psf_fwhm=self.psf_fwhm, aniso_param=aniso_param, num_evaluate=5000)
+                ll_vd = log_normal_pdf(inferred_vd,
+                                       self.measured_vd,
+                                       self.measured_vd_err)
+            # Time delays
+            inferred_td = lens_prop.time_delays(self.kwargs_lens, self.kwargs_ps, kappa_ext=k_ext)
+            if self.requires_reordering:
+                inferred_td = self.reorder_to_tdlmc(inferred_td)
+            else:
+                inferred_td = np.array(inferred_td)
+            inferred_td = inferred_td[1:] - inferred_td[0]
+            ll_td = np.sum(log_normal_pdf(inferred_td, self.measured_td, self.measured_td_err))
+            log_w = ll_vd + ll_td
+            w = mp.exp(log_w)
+
+            if n == 0 and np.isnan(float(w)):
+                warnings.warn("Weight is nan. Double check `abcd_ordering`.")
+
+            samples_dict['H0_samples'][n] = H0_candidate
+            samples_dict['H0_weights'][n] = w
+            if not exclude_vel_disp:
+                samples_dict['inferred_vd'][n] = inferred_vd
+                samples_dict['ll_vd'][n] = ll_vd
+            samples_dict['inferred_td'][n, :] = inferred_td
+            samples_dict['ll_td'][n] = ll_td
+
+            # Normalize weights to unity
+            samples_dict['H0_weights'] /= np.sum(samples_dict['H0_weights'])
```

Comment:
shouldn't this line be outside of the for loop? Currently you are re-normalizing after every proposal

Reply:
Thank you!